java/arcs/core/allocator/Allocator.kt

```diff
@@ -116,12 +116,33 @@ class Allocator(val hostRegistry: HostRegistry) {
         storageKey: CreateableStorageKey,
         type: Type
     ): StorageKey =
-        CapabilitiesResolver(CapabilitiesResolver.StorageKeyOptions(arcId)).createStorageKey(
-            storageKey.capabilities, type.toSchemaHash()
-        )?.childKeyForHandle(idGenerator.newChildId(arcId, "").toString())
-            ?: throw Exception(
-                "Unable to create storage key $storageKey"
+        CapabilitiesResolver(CapabilitiesResolver.CapabilitiesResolverOptions(arcId))
+            .createStorageKey(
+                storageKey.capabilities,
+                toEntitySchema(type),
+                idGenerator.newChildId(arcId, "").toString()
             )
+        ?: throw Exception(
+            "Unable to create storage key $storageKey"
+        )
+
+    /**
+     * Retrieves [Schema] from the given [Type], if possible.
+     * TODO: declare a common interface.
+     */
+    private fun toEntitySchema(type: Type): Schema {
+        when (type) {
+            is SingletonType<*> -> if (type.containedType is EntityType) {
+                return (type.containedType as EntityType).entitySchema
+            }
+            is CollectionType<*> -> if (type.collectionType is EntityType) {
+                return (type.collectionType as EntityType).entitySchema
+            }
+            is EntityType -> return type.entitySchema
+            else -> Unit
+        }
+        throw Exception("Can't retrieve entitySchema of unknown type $type")
```

Comment:
Nit: `IllegalArgumentException`, also you can throw this from the `else ->`, then lift the return. out of the other `when` branches.

Reply:
partially done (return inside `when` didn't work because of the extra `if` statements.. not sure if there is a better way..